### PR TITLE
implements ACLK env endpoint

### PR DESCRIPTION
**Pull request by underhood**
##### Summary
Implements getting data and parsing them from new `/env` endpoint. This will be used in later PRs

##### Component Name
ACLK

##### Test Plan
aclk_env should be properly populated by data from the response from the cloud. Can test with production cloud.
The agent should connect to the cloud successfully.

### CODE REVIEW

**Comment by stelfrag** [comment]
Looks like it connects and builds the env
```
(gdb) p **env->transports
$7 = {endpoint = 0x5555618f9210 "https://app.netdata.cloud/mqtt", type = ACLK_TRP_MQTT_3_1_1}
```

I will check the code now

File: aclk/aclk_otp.c [hunk 1]
```diff
@@ -253,9 +289,357 @@ void aclk_get_mqtt_otp(RSA *p_key, char *aclk_hostname, int port, char **mqtt_us
...
+    https_req_t req = HTTPS_REQ_T_INITIALIZER;
+    https_req_response_t resp = HTTPS_RES_RESPONSE_T_INITIALIZER;
+
+    req.request_type = HTTP_REQ_GET;
+
+    char *agent_id = is_agent_claimed();
+    if (agent_id == NULL)
+    {
+        error("Agent was not claimed - cannot perform challenge/response");
+        buffer_flush(buf);
```
Comment (stelfrag) [inline aclk/aclk_otp.c:610]: Did you mean to free the buffer in all those cases? Because you have buffer create in L599

Comment (underhood) [reply aclk/aclk_otp.c:610]: thx, you are right this is a mistake. I think originally buffer was given by caller function but then I changed it and forgot to update the flushes.... will fix this

Comment (underhood) [reply aclk/aclk_otp.c]: fixed

**Comment by underhood** [comment]
rebased after prereq PR merged to master

**Comment by stelfrag** [comment]
The agent will not connect or drops the connection. Trying to replicate reliably

**Comment by underhood** [comment]
@stelfrag if you are getting 409 that is the too many connection kick from cloud

**Comment by stelfrag** [comment]
> @stelfrag if you are getting 409 that is the too many connection kick from cloud

I managed to have two instances running -- scratch that!

**Comment by underhood** [comment]
> > @stelfrag if you are getting 409 that is the too many connection kick from cloud
> 
> I managed to have two instances running -- scratch that!

Yeah Up the stack we have the PR with TBEB and there we log nice error messages with cloud reasons etc.

File: aclk/aclk.c [hunk 2]
```diff
@@ -473,29 +494,67 @@ static int aclk_attempt_to_connect(mqtt_wss_client client)
...
+        aclk_get_mqtt_otp(aclk_private_key, &mqtt_otp_user, &mqtt_otp_pass, &auth_url);
+        url_t_destroy(&auth_url);
+
         mqtt_conn_params.clientid = mqtt_otp_user;
         mqtt_conn_params.username = mqtt_otp_user;
         mqtt_conn_params.password = mqtt_otp_pass;
 #endif
 
+        // Do the MQTT connection
+        ret = aclk_get_transport_idx(aclk_env);
```
Comment (stelfrag) [inline aclk/aclk.c:530]: This will now cause a crash if the ACLK_DISABLE_CHALLENGE is used

Comment (underhood) [reply aclk/aclk.c:530]: good catch forgot to retest the `ACLK_DISABLE_CHALLENGE`

Comment (underhood) [reply aclk/aclk.c:532]: should be fixed now, this is developer convenience flag

File: aclk/aclk_otp.c [hunk 3]
```diff
@@ -298,3 +298,335 @@ void aclk_get_mqtt_otp(RSA *p_key, char *aclk_hostname, int port, char **mqtt_us
...
+        return 2;
+    }
+    if (resp.http_code != 200) {
+        error("The HTTP code not 200 OK (Got %d)", resp.http_code);
+        https_req_response_free(&resp);
+        buffer_flush(buf);
+        return 3;
+    }
+
+    if (parse_json_env(resp.payload, &aclk_env)) {
```
Comment (vkalintiris) [inline aclk/aclk_otp.c]: I might be overtly worried but if the remote's response is malformed, is it possible to end up with a palyoad that is NULL?

Comment (underhood) [reply aclk/aclk_otp.c]: thats why we have this check few lines above:
`if (aclk_https_request(&req, &resp)) {`

Comment (underhood) [reply aclk/aclk_otp.c]: ah I see what you mean now.... you mean due to some cloud bug they could return `200 OK` but the payload will be empty. Added additional check. Thx

**Review by vkalintiris — APPROVED** [review]
LGTM. A couple of non-blocking comments inline.

**Review by stelfrag — APPROVED** [review]

**Comment by underhood** [comment]
@vkalintiris I will bring some of your comments in follow up (polish) PRs

**Pull request merged**